internal/provisioner/plain/controllers/bundleinstance_controller.go

```diff
@@ -262,12 +283,76 @@ func (r *BundleInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		Type:    rukpakv1alpha1.TypeInstalled,
 		Status:  metav1.ConditionTrue,
 		Reason:  rukpakv1alpha1.ReasonInstallationSucceeded,
-		Message: fmt.Sprintf("instantiated bundle %s successfully", bi.Spec.BundleName),
+		Message: fmt.Sprintf("instantiated bundle %s successfully", bundle.GetName()),
 	})
-	bi.Status.InstalledBundleName = bi.Spec.BundleName
+	bi.Status.InstalledBundleName = bundle.GetName()
+
+	if err := r.reconcileOldBundles(ctx, oldBundles, bundle); err != nil {
+		l.Error(err, "failed to delete old bundles")
+		return ctrl.Result{}, err
+	}
+
 	return ctrl.Result{}, nil
 }
 
+// reconcileDesiredBundle is responsible for checking whether the desired
+// Bundle resource that's specified in the BundleInstance parameter's
+// spec.Template configuration is present on cluster, and if not, creates
+// a new Bundle resource matching that desired specification.
+func reconcileDesiredBundle(ctx context.Context, c client.Client, bi *rukpakv1alpha1.BundleInstance) (*rukpakv1alpha1.Bundle, []*rukpakv1alpha1.Bundle, error) {
+	// get the set of Bundle resources that already exist on cluster, and sort
+	// by metadata.CreationTimestamp in the case there's multiple Bundles
+	// that match the label selector.
+	existingBundles, err := util.GetBundlesForBundleInstanceSelector(ctx, c, bi)
+	if err != nil {
+		return nil, nil, err
+	}
+	sort.Sort(util.BundlesByCreationTimestamp(existingBundles))
+
+	// check whether the BI controller has already reached the maximum
+	// generated Bundle limit to avoid hotlooping scenarios.
+	if len(existingBundles) > maxGeneratedBundleLimit {
+		return nil, nil, ErrMaxGeneratedLimit
+	}
+
+	// check whether there's an existing Bundle that matches the desired Bundle template
+	// specified in the BI resource, and if not, generate a new matches the template.
+	b := util.CheckExistingBundlesMatchesTemplate(existingBundles, bi.Spec.Template)
+	if b == nil {
+		controllerRef := metav1.NewControllerRef(bi, bi.GroupVersionKind())
+		b = &rukpakv1alpha1.Bundle{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            util.GenerateBundleName(bi.GetName()),
+				OwnerReferences: []metav1.OwnerReference{*controllerRef},
+				Labels:          bi.Spec.Template.Labels,
+			},
+			Spec: bi.Spec.Template.Spec,
+		}
+		if err := c.Create(ctx, b); err != nil {
+			return nil, nil, err
+		}
+	}
+	return b, existingBundles, err
```

Comment:
When we call this function, we call the second return value `oldBundles` in the calling scope, but here it looks like its all of the bundles that match the selector.  Should we go ahead of make sure that `b` is not present in `existingBundles`?

Or change the calling scope variable name to `existingBundles`?